libgalois/include/katana/EntityTypeManager.h

```diff
@@ -0,0 +1,227 @@
+/// A map from EntityTypeID to a set of EntityTypeIDs
+using EntityTypeIDToSetOfEntityTypeIDsMap = std::vector<SetOfEntityTypeIDs>;
+/// A map from the atomic type name to its EntityTypeID
+/// (that does not intersect any other atomic type)
+using AtomicTypeNameToEntityTypeIDMap =
+    std::unordered_map<std::string, EntityTypeID>;
+/// A map from the atomic type's EntityTypeID to its name
+/// (that does not intersect any other atomic type)
+using EntityTypeIDToAtomicTypeNameMap =
+    std::unordered_map<EntityTypeID, std::string>;
+
+class EntityTypeManager {
+public:
+  EntityTypeManager() { Init(); }
+
+  EntityTypeManager(
+      EntityTypeIDToAtomicTypeNameMap&& atomic_entity_type_id_to_type_name,
+      EntityTypeIDToSetOfEntityTypeIDsMap&&
+          entity_type_id_to_atomic_entity_type_ids)
+      : atomic_entity_type_id_to_type_name_(
+            std::move(atomic_entity_type_id_to_type_name)),
+        entity_type_id_to_atomic_entity_type_ids_(
+            std::move(entity_type_id_to_atomic_entity_type_ids)) {
+    for (auto& type_id_name_pair : atomic_entity_type_id_to_type_name_) {
+      atomic_type_name_to_entity_type_id_.emplace(
+          std::make_pair(type_id_name_pair.second, type_id_name_pair.first));
+    }
+
+    size_t num_entity_types = entity_type_id_to_atomic_entity_type_ids_.size();
+    atomic_entity_type_id_to_entity_type_ids_.resize(num_entity_types);
+    for (size_t i = 0, ni = num_entity_types; i < ni; ++i) {
+      for (size_t j = 0, nj = num_entity_types; j < nj; ++j) {
+        if (entity_type_id_to_atomic_entity_type_ids_[i].test(j)) {
+          atomic_entity_type_id_to_entity_type_ids_[j].set(i);
+        }
+      }
+    }
+  }
+
+  EntityTypeManager(
+      EntityTypeIDToAtomicTypeNameMap&& atomic_entity_type_id_to_type_name,
+      AtomicTypeNameToEntityTypeIDMap&& atomic_type_name_to_entity_type_id,
+      EntityTypeIDToSetOfEntityTypeIDsMap&&
+          entity_type_id_to_atomic_entity_type_ids,
+      EntityTypeIDToSetOfEntityTypeIDsMap&&
+          atomic_entity_type_id_to_entity_type_ids)
+      : atomic_entity_type_id_to_type_name_(
+            std::move(atomic_entity_type_id_to_type_name)),
+        atomic_type_name_to_entity_type_id_(
+            std::move(atomic_type_name_to_entity_type_id)),
+        entity_type_id_to_atomic_entity_type_ids_(
+            std::move(entity_type_id_to_atomic_entity_type_ids)),
+        atomic_entity_type_id_to_entity_type_ids_(
+            std::move(atomic_entity_type_id_to_entity_type_ids)) {}
+
+  void Reset() {
+    atomic_entity_type_id_to_type_name_.clear();
+    atomic_type_name_to_entity_type_id_.clear();
+    entity_type_id_to_atomic_entity_type_ids_.clear();
+    atomic_entity_type_id_to_entity_type_ids_.clear();
+    Init();
+  }
+
+  /// adds a new entity type for the atomic type with name \p name
+  /// \returns the EntityTypeID for the new type
+  EntityTypeID AddAtomicEntityType(const std::string& name) {
```

Comment:
This creates an atomic type with a specific name. I think the documentation is the main issue, but I want to make sure this is clear. There isn't really a distinction between atomic and composite entity types. They are all in the same "namespace" (ID space I guess) and have the same supported operations. So there is no real place where you create an entity type for an atomic type. You just create an atomic type which by definition is also an entity type.